File: resolver.c
#include <arpa/inet.h>
#include <errno.h>
#include <netinet/in.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/socket.h>
#include <string.h>
#include <sys/time.h>
#include <sys/types.h>
#include <unistd.h>

#include "dns.h"

#define MAX_QUERY_SIZE 1024
#define MAX_RESPONSE_SIZE 4096

// Note: uint8_t* is a pointer to 8 bits of data.

/**
 * Constructs a DNS query for hostname's Type A record.
 *
 * @param query Pointer to memory where query will stored.
 * @param hostname The host we are trying to resolve
 * @return The number of bytes in the constructed query.
 */
int construct_query(uint8_t* query, char* hostname) {
	memset(query, 0, MAX_QUERY_SIZE);

	// first part of the query is a fixed size header
	DNSHeader *hdr = (DNSHeader*)query;

	// set ID to 5... you should randomize this!
	hdr->id = htons(5);

	// set header flags to request iterative query
	hdr->flags = htons(0x0000);	

	// 1 question, no answers or other records
	hdr->q_count=htons(1);
	hdr->a_count=htons(0);
	hdr->auth_count=htons(0);
	hdr->other_count=htons(0);

	// We are going to have to wade into pointer arithmetic here since our
	// struct is a fixed size but our queries will be variably sized.

	// add the name
	int query_len = sizeof(DNSHeader); 
	int name_len = convertStringToDNS(hostname,query+query_len);
	query_len += name_len; 
	
	// set the query type to A (i.e. 1)
	uint16_t *type = (uint16_t*)(query+query_len);
	*type = htons(1);
	query_len+=2;

	// finally the class: INET
	uint16_t *class = (uint16_t*)(query+query_len);
	*class = htons(1);
	query_len += 2;
 
	return query_len;
}


int main(int argc, char **argv) {
	if (argc < 2) {
		printf("usage: %s [-m] hostname\n", argv[0]);
		exit(1);
	}

	int sock = socket(AF_INET, SOCK_DGRAM, 0);
	if (sock < 0) {
		perror("socket");
		exit(0);
	}

	/* Create a time value structure and set it to five seconds. */
	struct timeval tv;
	memset(&tv, 0, sizeof(struct timeval));
	tv.tv_sec = 5;

	/* Tell the OS to use that time value as a time out for operations on
	 * our socket. */
	int res = setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, &tv,
			sizeof(struct timeval));

	if (res < 0) {
		perror("setsockopt");
		exit(0);
	}

	in_addr_t nameserver_addr = inet_addr("172.16.7.15");
	
	struct sockaddr_in addr; 	// internet socket address data structure
	addr.sin_family = AF_INET;
	addr.sin_port = htons(53); // port 53 for DNS
	addr.sin_addr.s_addr = nameserver_addr; // destination address (any local for now)

	// uint8_t is a standard, unsigned 8-bit value.
	// You should use that type for all buffers used for sending to and
	// receiving from the DNS server.
	uint8_t query[MAX_QUERY_SIZE]; 
	int query_len=construct_query(query, "www.sandiego.edu");

	int send_count = sendto(sock, query, query_len, 0,
							(struct sockaddr*)&addr, sizeof(addr));
	if(send_count<0) { perror("Send failed");	exit(1); }

	socklen_t len = sizeof(struct sockaddr_in);

	uint8_t response[MAX_RESPONSE_SIZE];

	/* Blocking calls will now return error (-1) after the timeout period with
	 * errno set to EAGAIN. */
	res = recvfrom(sock, response, MAX_RESPONSE_SIZE, 0, 
					(struct sockaddr *)&addr, &len);

	if (res < 1) {
		if (errno == EAGAIN) {
			printf("Timed out!\n");
		} else {
			perror("recv");
		}
	}

	return 0;
}

File: Resolver.java
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.net.DatagramSocket;
import java.net.DatagramPacket;
import java.net.InetAddress;
import java.net.SocketException;
import java.net.UnknownHostException;
import java.net.SocketTimeoutException;

public class Resolver {

	/**
	 * Converts our Java string (e.g. www.sandiego.edu) to the DNS-style
	 * string (e.g. (3)www(8)sandiego(3)edu(0).
	 *
	 * @param s The string you want to convert.
	 * @returns Buffer containing the converted string.
	 */
	public static ByteBuffer stringToNetwork(String s) {
		ByteBuffer bb = ByteBuffer.allocate(s.length()+2);
		bb.order(ByteOrder.BIG_ENDIAN);

		String[] parts = s.split("\\.");
		for (String p : parts) {
			bb.put((byte)p.length());
			bb.put(p.getBytes());
		}

		bb.put((byte)0);

		return bb;
	}

	/**
	 * Converts a DNS-style string (e.g. (3)www(8)sandiego(3)edu(0) to the normal
	 * Java (e.g. www.sandiego.edu).
	 * This supports DNS-style name compression (see RFC 1035, Section 4.1.4).
	 *
	 * @param message The entire DNS message, in ByteBuffer format.
	 * @param start The location in the message buffer where the network
	 * 				string starts.
	 * @param sb StringBuilder to construct the string as we are reading the
	 * 				message.
	 * @return Location in message immediately after the string.
	 */
	private static int getStringFromDNS(ByteBuffer message, int start, StringBuilder sb) {
		int position = start;

		byte currByte = message.get(position);
		position += 1;

		while (currByte != 0) {
			byte length = currByte;
			byte wtf = (byte)((length >> 6) & 3);

			// If length starts with "11" then this is a compressed name
			//if (wtf == 3) {
			if ((length >> 6 & 3) == 3) {
				// next byte will contain rest of pointer
				byte nextByte = message.get(position); 
				int offset = (length & 0x3F) << 8;
				offset += nextByte;

				getStringFromDNS(message, offset, sb);
				return position + 1;
			}

			for (int i = 0; i < length; i++) {
				byte c = message.get(position);
				sb.append((char)c);
				position += 1;
			}
			sb.append('.');

			currByte = message.get(position);
			position += 1;
		}

		sb.deleteCharAt(sb.length()-1);

		return position;
	}

	/**
	 * Constructs a DNS query for hostname's Type A record.
	 *
	 * @param hostname The host we are trying to resolve
	 * @return A ByteBuffer with the query
	 */
	public static ByteBuffer constructQuery(String hostname) {
		ByteBuffer bb = ByteBuffer.allocate(256);

		// Network order is BIG_ENDIAN (opposite of what x86 uses)
		bb.order(ByteOrder.BIG_ENDIAN);

		bb.putShort((short)5); // id set to 5... should be random!
		bb.putShort((short)0); // flags (set to make an iterative request)

		// 1 question, no answers, no auth, or other records
		bb.putShort((short)1); // num questions
		bb.putShort((short)0); // num answers
		bb.putShort((short)0); // num auth
		bb.putShort((short)0); // num other

		// Create and add the host name
		ByteBuffer nameString = Resolver.stringToNetwork(hostname);
		bb.put(nameString.array(), 0, nameString.position());
		bb.putShort((short)1); // query type == 1 (i.e. Type A)
		bb.putShort((short)1); // class: INET

		return bb;
	}

	public static void main(String[] args) throws Exception {
		DatagramSocket sock = new DatagramSocket();
		sock.setSoTimeout(5000); // set timeout to 5000ms (i.e. 5 seconds)

		// construct a query and set that request
		ByteBuffer bb = constructQuery("www.sandiego.edu");

		String serverIP = "172.16.7.15"; // This is USD's local DNS server.
										 // You should NOT use it in your final
										 // program.

		// Using InetAddress.getByName in this project is only allowed if the
		// input is an IP address (i.e. don't give it "www.foo.com")
		DatagramPacket request = new DatagramPacket(bb.array(), 
													bb.position(),
													InetAddress.getByName(serverIP), 
													53);
		sock.send(request);

		ByteBuffer response_bb = ByteBuffer.allocate(256);

		DatagramPacket response = new DatagramPacket(response_bb.array(),
														response_bb.capacity());
		try {
			sock.receive(response);
		} catch (SocketTimeoutException ste) {
			System.out.println("Timed out!");
		} catch (Exception e) {
			System.out.println("Error receiving: " + e);
		}

		// Now you'll need to interpret the response by reading the
		// response_bb and interpretting the values you find there.
	}
}
